docs(api): add pages for NormalModuleFactory and ContextModuleFactory Hooks (#4481)
webpack/webpack.js.org: src/content/api/contextmodulefactory-hooks.md

---
title: ContextModuleFactory Hooks
group: Plugins
sort: 11
contributors:
  - iguessitsokay
---

The `ContextModuleFactory` module is used by the `Compiler` to generate dependencies from webpack specific [require.context](/api/module-methods/#requirecontext) API. It resolves the requested directory, generates requests for each file and filters against passed regExp. Matching dependencies then passes through [NormalModuleFactory](/api/normalmodulefactory-hooks).

The `ContextModuleFactory` class extends `Tapable` and provides the following
lifecycle hooks. They can be tapped the same way as compiler hooks:

``` js
ContextModuleFactory.hooks.someHook.tap(/* ... */);
```

As with the `compiler`, `tapAsync` and `tapPromise` may also be available
depending on the type of hook.

> ### `beforeResolve`

Reviewer: ```suggestion
### beforeResolve
```
Let's remove inline code style from heading as it's sort of disturbing when applied on headings. So far there're many usages on this site, but I'm going to remove them in the future.

`AsyncSeriesWaterfallHook`

> Fired before resolving the requested directory. The request can be ignored by returning `false`.

Reviewer: ```suggestion
Called before resolving the requested directory. The request can be ignored by returning `false`.
```
`Fired` is good in my opinion, unfortunately the linter doesn't think so, thus let's use `called` here.

- Callback Parameters: `data`

> ### `afterResolve`

Reviewer: ```suggestion
### afterResolve
```

`AsyncSeriesWaterfallHook`

> Fired after resolving the requested directory.

Reviewer: ```suggestion
Called after the requested directory resolved.
```

- Callback Parameters: `data`

> ### `contextModuleFiles`

Reviewer: ```suggestion
### contextModuleFiles
```

`SyncWaterfallHook`

> Fired after directory contents are read. On recursive mode, fires for each sub-directory as well. Callback parameter is an array of all file and folder names in each directory.

Reviewer: ```suggestion
Called after directory contents are read. On recursive mode, calls for each sub-directory as well. Callback parameter is an array of all file and folder names in each directory.
```

- Callback Parameters: `fileNames`

> ### `alternativeRequests`

Reviewer: ```suggestion
### alternativeRequests
```

`AsyncSeriesWaterfallHook`

> Fired for each file after the request is created but before filtering against regExp.

Reviewer: ```suggestion
Called for each file after the request is created but before filtering against regExp.
```
